src/server/detail/save_stages_controller.h

```diff
@@ -18,6 +18,37 @@ class Transaction;
 class Service;
 
 namespace detail {
+
+class SnapshotStorage {
+ public:
+  virtual ~SnapshotStorage() = default;
+
+  virtual GenericError OpenFile(const std::string& path, io::Sink** file, bool* is_direct,
+                                bool* is_linux_file) = 0;
+};
+
```

Comment:
Cool. If they grow I'd suggest moving it a separate file. 

I wouldn't use so much output arguments. The whole problem is that we really don't have a proper std::expected wrapper that we commonly use and which is convenient. The options are our custom unexpected and just an std::variant. (Added this as a todo for me)

I also assume something below uses const string&, so string_view is not an option, right?

Other than that, maybe it would make sense to replace the two boolean variables with an enum, like `direct_file, iouring_file, epoll_file, cloud_aws`... and in the future cloud_gcp. So the cloud implementations don't have to care about those even only in their interface

Reply:
> I wouldn't use so much output arguments

Sure - in this case I guess could use `io::Result<pair<io::Sink, FileType>>` (if using a file type enum instead of booleans)?

> I also assume something below uses const string&, so string_view is not an option, right?

Yep

> replace the two boolean variables with an enum, like direct_file, iouring_file, epoll_file, cloud_aws

Ok sure - so a 'file type' enum, then the type is a bitmask of the different values? eg `direct_file | linux_file`? (otherwise without a bitmask i guess you'd end up a bunch of complex comparisons like `if type == io_uring_file || type == epoll_file` to work out if its a linux file)

Reply:
> io::Result<pair<io::Sink, FileType>>

Yes, but I'd keep the generic error because it describes the issue more in detail. Just Io::Result has error_code as its base, so you need `io::Result<pair, GenericError>`

True, without the bitmask you can end up with different complicated conidtions. Currently `direct => value == direct_file`, `file => value <= epoll`.

Masking is also an option, it seems even easier. So we'll have for now 3 options: direct, file, cloud

Reply:
Ok sure thanks - updated